lightning/src/ln/onchaintx.rs

```diff
@@ -709,4 +778,47 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
 			self.pending_claim_requests.remove(&req);
 		}
 	}
+
+	pub(super) fn provide_latest_local_tx(&mut self, tx: LocalCommitmentTransaction) -> Result<(), ()> {
+		// To prevent any unsafe state discrepancy between offchain and onchain, once local
+		// commitment transaction has been signed due to an event (either block height for
+		// HTLC-timeout or channel force-closure), don't allow any further update of local
+		// commitment transaction view to avoid delivery of revocation secret to counterparty
+		// for the aformentionned signed transaction.
+		if let Some(ref local_commitment) = self.local_commitment {
+			if local_commitment.has_local_sig() { return Err(()) }
+		}
+		self.prev_local_commitment = self.local_commitment.take();
+		self.local_commitment = Some(tx);
+		Ok(())
+	}
+
+	pub(super) fn get_fully_signed_local_tx(&mut self, channel_value_satoshis: u64) -> Option<Transaction> {
+		if let Some(ref mut local_commitment) = self.local_commitment {
+			self.key_storage.sign_local_commitment(local_commitment, &self.funding_redeemscript, channel_value_satoshis, &self.secp_ctx);
+			return Some(local_commitment.with_valid_witness().clone());
+		}
+		None
+	}
+
+	#[cfg(test)]
+	pub(super) fn get_fully_signed_copy_local_tx(&mut self, channel_value_satoshis: u64) -> Option<Transaction> {
+		if let Some(ref mut local_commitment) = self.local_commitment {
+			let mut local_commitment = local_commitment.clone();
```

Comment:
The thing that should be enforcing that we only ever sign one version is the enforcing_trait_impls stuff, not here. We need a #[cfg(test)] different version of `sign_local_commitment`.

Reply:
After further analysis, this test-only function doesn't make sense at all, `sign_local_commitment`  only enforce uniqueness of signature for a commitment transaction version (`has_local_sig`).  For testing purposes what we want is a) signature of multiple different commitment transactions b) avoid Monitor panic/error after local-commitment-state-lockdown. So modified `provide_latest_local_tx` to not return err for test-only.

Reply:
Right, we don't currently enforce anything, we *will* do it at the ChannelKeys layer. Thus, we need a test-only version of sign_local_comitment.

Reply:
Updated version in enforcing_trait_impls, with a TODO in keysinterface